### PR TITLE
refactor: add generics to ClientStorage

### PR DESCRIPTION
**Pull request by beyackle**
## Description

This adds a generic to our ClientStorage class so we have more type-safety about what we put into and get out of there (which lets us use fewer "any" type annotations).

## Task Item

#minor

### CODE REVIEW

**Comment by beyackle** [comment]
There's a bug here; a `ClientStorage<string>` will try to deserialize a string as if it were JSON and crash instead of just spitting the string out. The older version handled this using a slightly clumsy mechanism, so I want to figure out a better way to do this.

**Comment by beyackle** [comment]
No, actually, string storage will work. JSON.stringify sticks an extra pair of quotes around strings when they serialize (so `abc` becomes `"abc"`), and then JSON.parse will turn that back into a regular string as intended.

File: Composer/packages/client/src/utils/luFileStatusStorage.ts [hunk 1]
```diff
@@ -6,14 +6,16 @@ import storage, { ClientStorage } from './storage';
 
 const KEY = 'LuFileStatus';
 
-interface ILuFileStatus {
+type LuFileStatus = {
   [fileId: string]: boolean;
-}
+};
+
+type LuFileRecord = { [projectId: string]: LuFileStatus };
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/utils/luFileStatusStorage.ts:13]: These can be converted to `Record<string, *>`

File: Composer/packages/client/src/utils/qnaFileStatusStorage.ts [hunk 2]
```diff
@@ -6,14 +6,16 @@ import storage, { ClientStorage } from './storage';
 
 const KEY = 'QnaFileStatus';
 
-interface IQnaFileStatus {
+type QnaFileStatus = {
   [fileId: string]: boolean;
-}
+};
+
+type QnaFileRecord = { [projectId: string]: QnaFileStatus };
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/utils/qnaFileStatusStorage.ts:13]: Use `Record` here as well.

File: Composer/packages/client/src/utils/storage.ts [hunk 3]
```diff
@@ -12,15 +12,17 @@ export class ClientStorage {
...
       return this.remove(this.prefix(key));
     }
     this.storage.setItem(this.prefix(key), this.serialize(val));
     return val;
   }
 
-  public get<T = any>(key: string, def?: T): T {
+  public get(key: string): T | undefined;
+  public get(key: string, def: T): T;
+  public get(key: string, def?: T): T | undefined {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/utils/storage.ts:25]: Can we take this opportunity to rename this argument. Something like `defaultValue` would be better than `def` which is ambiguous.

File: Composer/packages/client/src/utils/storage.ts [hunk 4]
```diff
@@ -37,36 +39,37 @@ export class ClientStorage {
     this.storage.clear();
   }
 
-  public getAll(): { [key: string]: any } {
+  public getAll(): { [key: string]: TextDecoderOptions } {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/utils/storage.ts:42]: Consider the `Record` type here.

**Review by a-b-r-o-w-n — CHANGES_REQUESTED** [review]

**Comment by a-b-r-o-w-n** [comment]
Stale - closing. cc @beyackle

**Pull request closed without merging**